dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -1566,4 +1568,37 @@ private void checkNotHosted(Workflow workflow) {
             throw new WebApplicationException("Cannot modify hosted entries this way", HttpStatus.SC_BAD_REQUEST);
         }
     }
+
+    @GET
+    @Timed
+    @UnitOfWork
+    @Path("/{workflowId}/zip/{workflowVersionId}")
+    @ApiOperation(value = "Download a ZIP file of a workflow and all associated files.", authorizations = { @Authorization(value = JWT_SECURITY_DEFINITION_NAME) })
+    @Produces("application/zip")
+    public Response getWorkflowZip(@ApiParam(hidden = true) @Auth Optional<User> user, @ApiParam(value = "workflowId", required = true) @PathParam("workflowId") Long workflowId,
+                                   @ApiParam(value = "workflowVersionId", required = true) @PathParam("workflowVersionId") Long workflowVersionId) {
+        Workflow workflow;
+        if (user.isPresent()) {
+            workflow = workflowDAO.findById(workflowId);
+            checkEntry(workflow);
+            checkCanReadWorkflow(user.get(), workflow);
+        } else {
+            workflow = workflowDAO.findPublishedById(workflowId);
+            checkEntry(workflow);
+        }
+
+        WorkflowVersion workflowVersion = getWorkflowVersion(workflow, workflowVersionId);
+        Set<SourceFile> sourceFiles = workflowVersion.getSourceFiles();
+        if (sourceFiles == null || sourceFiles.size() == 0) {
+            return null;
```

Comment:
What does `return null` do -- what code will be returned?

I wonder if it would be better to explicitly return an HTTP status code, although I don't know if this would be a 5xx or a 4xx.

Reply:
Probably nothing good